libs/langchain/langchain/chat_models/fireworks.py

```diff
@@ -0,0 +1,140 @@
+import os
+from typing import Any, Dict, Iterator, List, Mapping, Optional, Tuple
+from langchain.adapters.openai import convert_dict_to_message, convert_message_to_dict
+
+from langchain.callbacks.manager import (
+    CallbackManagerForLLMRun,
+)
+from langchain.chat_models.base import BaseChatModel
+from langchain.schema.messages import (
+    AIMessageChunk,
+    BaseMessage,
+    BaseMessageChunk,
+    ChatMessageChunk,
+    FunctionMessageChunk,
+    HumanMessageChunk,
+    SystemMessageChunk,
+)
+from langchain.schema.output import ChatGeneration, ChatGenerationChunk, ChatResult
+import openai
+
+
+def _convert_delta_to_message_chunk(
+    _dict: Mapping[str, Any], default_class: type[BaseMessageChunk]
+) -> BaseMessageChunk:
+    role = _dict.get("role")
+    content = _dict.get("content") or ""
+    if _dict.get("function_call"):
+        additional_kwargs = {"function_call": dict(_dict["function_call"])}
+    else:
+        additional_kwargs = {}
+
+    if role == "user" or default_class == HumanMessageChunk:
+        return HumanMessageChunk(content=content)
+    elif role == "assistant" or default_class == AIMessageChunk:
+        return AIMessageChunk(content=content, additional_kwargs=additional_kwargs)
+    elif role == "system" or default_class == SystemMessageChunk:
+        return SystemMessageChunk(content=content)
+    elif role == "function" or default_class == FunctionMessageChunk:
+        return FunctionMessageChunk(content=content, name=_dict["name"])
+    elif role or default_class == ChatMessageChunk:
+        return ChatMessageChunk(content=content, role=role)
+    else:
+        return default_class(content=content)
+
+
+class ChatFireworks(BaseChatModel):
```

Comment:
We should probably add completion with retrying.

Reply:
openai create api has retrying itself: https://github.com/openai/openai-python/blob/5d50e9e3b39540af782ca24e65c290343d86e1a9/openai/api_resources/chat_completion.py#L23

Reply:
sg. Do we throw the correct exception to trigger the retries? It's a bit wired that their langchain code contains an additional wrapper: https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/llms/openai.py#L103

I'm fine with you code.

Reply:
oops, openai might only retry when rcode=409
https://github.com/openai/openai-python/blob/e389823ba013a24b4c32ce38fa0bd87e6bccae94/openai/api_requestor.py#L448

I will make sure adding retry.

Reply:
I changed to use our fireworks python client instead of using openai's because our error_handling is a little different from openai, and it is more accurate to use ours in terms of error handling. We further need it to decide if this error type should continue retries.